nullaway/src/main/java/com/uber/nullaway/ClassAnnotationInfo.java

```diff
@@ -0,0 +1,194 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.util.Context;
+
+/**
+ * Provides APIs for querying whether code is annotated for nullness checking, and for related
+ * queries on what annotations are present on a class and/or enclosing classes. Makes use of caching
+ * internally for performance.
+ */
+public final class ClassAnnotationInfo {
+
+  private static final Context.Key<ClassAnnotationInfo> ANNOTATION_INFO_KEY = new Context.Key<>();
+
+  private static final int MAX_CACHE_SIZE = 200;
+
+  private final Cache<Symbol.ClassSymbol, CacheRecord> cache =
+      CacheBuilder.newBuilder().maximumSize(MAX_CACHE_SIZE).build();
+
+  private ClassAnnotationInfo() {}
+
+  /**
+   * Get the ClassAnnotationInfo for the given javac context. We ensure there is one instance per
+   * context (as opposed to using static fields) to avoid memory leaks.
+   */
+  public static ClassAnnotationInfo instance(Context context) {
+    ClassAnnotationInfo cache = context.get(ANNOTATION_INFO_KEY);
+    if (cache == null) {
+      cache = new ClassAnnotationInfo();
+      context.put(ANNOTATION_INFO_KEY, cache);
+    }
+    return cache;
+  }
+
+  /**
+   * Checks if a symbol comes from an annotated package, as determined by either configuration flags
+   * (e.g. {@code -XepOpt:NullAway::AnnotatedPackages}) or package level annotations (e.g. {@code
+   * org.jspecify.nullness.NullMarked}).
+   *
+   * @param outermostClassSymbol symbol for class (must be an outermost class)
+   * @param config NullAway config
+   * @return true if the class is from a package that should be treated as properly annotated
+   *     according to our convention (every possibly null parameter / return / field
+   *     annotated @Nullable), false otherwise
+   */
+  private static boolean fromAnnotatedPackage(
+      Symbol.ClassSymbol outermostClassSymbol, Config config) {
+    final String className = outermostClassSymbol.getQualifiedName().toString();
+    if (!config.fromExplicitlyAnnotatedPackage(className)
+        && !ASTHelpers.hasDirectAnnotationWithSimpleName(
+            outermostClassSymbol.packge(), NullabilityUtil.NULLMARKED_SIMPLE_NAME)) {
+      // By default, unknown code is unannotated unless @NullMarked or configured as annotated by
+      // package name
+      return false;
+    }
+    if (config.fromExplicitlyUnannotatedPackage(className)) {
+      // Any code explicitly marked as unannotated in our configuration is unannotated, no matter
+      // what.
+      return false;
+    }
+    if (config.shouldTreatGeneratedAsUnannotated()
+        && ASTHelpers.hasDirectAnnotationWithSimpleName(outermostClassSymbol, "Generated")) {
```

Comment:
Not sure if it's worth refactoring this part further at this point, but do note that both before and after this PR, this is technically looking at annotations at the class level for a method that presumably is asking only if the class comes from an annotated package...

Reply:
Went ahead and fixed this one too.  9dfdeed857ac41f2d8432af92155e1f6348ccdc9